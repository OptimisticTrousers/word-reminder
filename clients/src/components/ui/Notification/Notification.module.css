.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --text-color: var(--success-text-color);
  --icon-color: var(--success-icon-color);
  --button-bkg-color: var(--success-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-success-button-bkg-color);
  background-color: var(--success-bkg-color);
}

.notification--error {
  --text-color: var(--alert-text-color);
  --icon-color: var(--alert-icon-color);
  --button-bkg-color: var(--alert-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-alert-button-bkg-color);
  background-color: var(--alert-bkg-color);
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--text-color);
  padding: 0.375rem;
  background-color: var(--button-bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__button:hover {
  background-color: var(--secondary-button-bkg-color);
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
