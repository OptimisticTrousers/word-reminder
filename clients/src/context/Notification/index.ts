export { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
export { NotificationProvider } from "./Provider";
