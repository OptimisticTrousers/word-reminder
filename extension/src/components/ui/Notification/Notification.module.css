.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
