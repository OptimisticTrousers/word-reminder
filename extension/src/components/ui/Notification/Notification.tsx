import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
