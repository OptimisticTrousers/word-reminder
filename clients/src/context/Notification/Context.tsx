import { createContext } from "react";

import { NotificationContext as Context } from "./Provider";

export enum NOTIFICATION_ACTIONS {
  SUCCESS = "success",
  ERROR = "error",
}

export const NotificationContext = createContext<Context>({
  showNotification: function () {},
  dismissNotification: function () {},
});
