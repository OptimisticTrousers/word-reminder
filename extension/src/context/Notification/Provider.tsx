import { useReducer, ReactNode } from "react";

import { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
import { Notification } from "../../components/ui/Notification/Notification";

interface Props {
  children: ReactNode;
}

interface NotificationState {
  isVisible: boolean;
  type: string;
  message: string;
}

interface NotificationAction {
  payload: { message: string };
  type: string;
}

export interface NotificationContext {
  showNotification: (type: string, message: string) => void;
  dismissNotification: () => void;
}

const notificationReducer = (
  state: NotificationState,
  action: NotificationAction
) => {
  switch (action.type) {
    case NOTIFICATION_ACTIONS.SUCCESS:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.SUCCESS,
        message: action.payload.message,
      };
    case NOTIFICATION_ACTIONS.ERROR:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.ERROR,
        message: action.payload.message,
      };
    case DISAPPEAR:
      return { ...state, isVisible: false, type: "", message: "" };
    default: {
      throw Error("Unknown action: " + action.type);
    }
  }
};

// Provide ability to spawn Toast notification globally within the application
export function NotificationProvider({ children }: Props) {
  const [state, dispatch] = useReducer(notificationReducer, {
    isVisible: false,
    type: "",
    message: "",
  });

  function dismissNotification() {
    dispatch({
      type: DISAPPEAR,
      payload: { message: "" },
    });
  }

  function showNotification(type: string, message: string): void {
    const delay = 5000;

    dispatch({ type, payload: { message } });

    setTimeout(() => {
      dismissNotification();
    }, delay);
  }

  const { isVisible, type, message } = state;

  return (
    <NotificationContext.Provider
      value={{ showNotification, dismissNotification }}
    >
      {isVisible && <Notification type={type} message={message} />}
      {children}
    </NotificationContext.Provider>
  );
}

const DISAPPEAR = "disappear";
