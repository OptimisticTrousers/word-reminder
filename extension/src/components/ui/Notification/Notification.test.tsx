import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import { Notification } from "./Notification";

describe("Notification component", () => {
  it("renders notification with success class", async () => {
    const mockDismissNotification = vi.fn();
    const props = {
      type: NOTIFICATION_ACTIONS.SUCCESS,
      message: "Success Message.",
      dismissNotification: mockDismissNotification,
    };
    const user = userEvent.setup();

    const { asFragment } = render(
      <NotificationContext.Provider
        value={{
          dismissNotification: mockDismissNotification,
          showNotification: vi.fn(),
        }}
      >
        <Notification {...props} />
      </NotificationContext.Provider>
    );
    const dismissButton = screen.getByRole("button", { name: "Dismiss" });
    await user.click(dismissButton);

    const notification = screen.getByRole("dialog");
    const message = screen.getByText(props.message);
    expect(mockDismissNotification).toHaveBeenCalledTimes(1);
    expect(mockDismissNotification).toHaveBeenCalledWith();
    expect(message).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
    expect(notification.getAttribute("class")).toContain(
      `notification--${NOTIFICATION_ACTIONS.SUCCESS}`
    );
  });

  it("renders notification with error class", async () => {
    const mockDismissNotification = vi.fn();
    const props = {
      type: NOTIFICATION_ACTIONS.ERROR,
      message: "Error Message.",
      dismissNotification: mockDismissNotification,
    };
    const user = userEvent.setup();

    const { asFragment } = render(
      <NotificationContext.Provider
        value={{
          dismissNotification: mockDismissNotification,
          showNotification: vi.fn(),
        }}
      >
        <Notification {...props} />
      </NotificationContext.Provider>
    );
    const dismissButton = screen.getByRole("button", { name: "Dismiss" });
    await user.click(dismissButton);

    const notification = screen.getByRole("dialog");
    const message = screen.getByText(props.message);
    expect(mockDismissNotification).toHaveBeenCalledTimes(1);
    expect(mockDismissNotification).toHaveBeenCalledWith();
    expect(message).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
    expect(notification.getAttribute("class")).toContain(
      `notification--${NOTIFICATION_ACTIONS.ERROR}`
    );
  });
});
