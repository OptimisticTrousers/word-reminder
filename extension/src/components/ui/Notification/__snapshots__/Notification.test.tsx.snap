// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notification component > renders notification with error class 1`] = `[Function]`;

exports[`Notification component > renders notification with success class 1`] = `[Function]`;
